pkg/inspector/image-inspector.go

```diff
@@ -107,152 +116,108 @@ func NewDefaultImageInspector(opts iicmd.ImageInspectorOptions) ImageInspector {
 		}
 		inspector.imageServer = apiserver.NewWebdavImageServer(imageServerOpts)
 	}
+
+	inspector.scanOutputs.ScanResults = iiapi.ScanResult{
+		APIVersion: iiapi.DefaultResultsAPIVersion,
+		ImageName:  inspector.opts.Image,
+		Results:    []iiapi.Result{},
+	}
+
 	return inspector
 }
 
+type DockerRuntimeClient interface {
+	InspectImage(name string) (*docker.Image, error)
+	ContainerChanges(id string) ([]docker.Change, error)
+	PullImage(opts docker.PullImageOptions, auth docker.AuthConfiguration) error
+	CreateContainer(opts docker.CreateContainerOptions) (*docker.Container, error)
+	RemoveContainer(opts docker.RemoveContainerOptions) error
+	InspectContainer(id string) (*docker.Container, error)
+	DownloadFromContainer(id string, opts docker.DownloadFromContainerOptions) error
+}
+
 // Inspect inspects and serves the image based on the ImageInspectorOptions.
 func (i *defaultImageInspector) Inspect() error {
+	err := i.acquireAndScan()
+	if err != nil {
+		i.meta.ImageAcquireError = err.Error()
+	}
+
+	if i.imageServer != nil {
+		return i.imageServer.ServeImage(&i.meta, i.opts.DstPath, i.scanOutputs.ScanResults, i.scanOutputs.ScanReport, i.scanOutputs.HtmlScanReport)
+	}
+
+	return err
+}
+
+// AcquireAndScan acquires and scans the image based on the ImageInspectorOptions.
+func (i *defaultImageInspector) acquireAndScan() error {
 	var (
 		scanner iiapi.Scanner
 		err     error
 
-		scanReport, htmlScanReport []byte
-		filterFn                   iiapi.FilesFilter
+		filterFn iiapi.FilesFilter
 	)
 
-	scanResults := iiapi.ScanResult{
-		APIVersion: iiapi.DefaultResultsAPIVersion,
-		ImageName:  i.opts.Image,
-		Results:    []iiapi.Result{},
-	}
+	ctx := context.Background()
 
 	client, err := docker.NewClient(i.opts.URI)
 	if err != nil {
-		return fmt.Errorf("Unable to connect to docker daemon: %v\n", err)
-	}
-
-	ctx := context.Background()
-
-	if len(i.opts.Container) == 0 {
-		imageMetaBefore, inspectErrBefore := client.InspectImage(i.opts.Image)
-		if i.opts.PullPolicy == iiapi.PullNever && inspectErrBefore != nil {
-			return fmt.Errorf("Image %s is not available and pull-policy %s doesn't allow pulling",
-				i.opts.Image, i.opts.PullPolicy)
-		}
-
-		if i.opts.PullPolicy == iiapi.PullAlways ||
-			(i.opts.PullPolicy == iiapi.PullIfNotPresent && inspectErrBefore != nil) {
-			if err = i.pullImage(client); err != nil {
-				return err
-			}
-		}
-
-		imageMetaAfter, inspectErrAfter := client.InspectImage(i.opts.Image)
-		if inspectErrBefore == nil && inspectErrAfter == nil &&
-			imageMetaBefore.ID == imageMetaAfter.ID {
-			log.Printf("Image %s was already available", i.opts.Image)
-		}
-
-		randomName, err := generateRandomName()
-		if err != nil {
-			return err
-		}
-
-		imageMetadata, err := i.createAndExtractImage(client, randomName)
-		if err != nil {
-			return err
-		}
-		i.meta.Image = *imageMetadata
-		scanResults.ImageID = i.meta.Image.ID
+		i.meta.ImageAcquireError = err.Error()
```

Comment:
@enoodle please keep the logic of this (already too long) method simple and return on error.

If you need to serve even on error please use a wrapper method,